Allow compilation with minimal OpenSSL

cotequeiroz opened the pull request:
Updated use of OpenSSL deprecated API, so that Squid can be compiled
with OpenSSL built with the OPENSSL_NO_DEPRECATED option. Such OpenSSL
builds are useful for saving storage space on embedded systems.

Also added compat/openssl.h -- a centralized OpenSSL portability shim.
Including it is now required before #including openssl/*.h headers.

yadij commented on src/security/Session.h:
```diff
@@ -18,6 +18,31 @@
 #if HAVE_OPENSSL_SSL_H
 #include <openssl/ssl.h>
+
+/// These are not included by openssl/ssl.h when deprecated API is disabled
```
These includes should be added to the .h or .cc where the symbols they define are needed. Prefer the src/security/forward.h if there is a spread of usage.

yadij commented on src/security/Session.h:
```diff
@@ -18,6 +18,31 @@
+
+/// Use new calls for API deprecated in OpenSSL 1.1.0
+/// Note that all of the calls in this next group were implemented at the same time
+/// as X509_STORE_CTX_set0_untrusted, in both OpenSSL 1.1.0, and LibreSSL 2.7.0.
```
This statement is not completely true. At least LibreSSL 2.7.4 still contains SSLeay_version() symbols.

yadij commented on src/security/Session.h:
```diff
@@ -18,6 +18,31 @@
+/// Note that all of the calls in this next group were implemented at the same time
+/// as X509_STORE_CTX_set0_untrusted, in both OpenSSL 1.1.0, and LibreSSL 2.7.0.
+#if !HAVE_LIBCRYPTO_X509_STORE_CTX_SET0_UNTRUSTED
+#define OPENSSL_VERSION SSLEAY_VERSION
```
NP: OpenSSL version and SSLEAY version are not necessarily the same thing.

If (as I suspect) the removal of SSLEAY version lookup is related to the relicensing of OpenSSL. Then the restricted builds may have different license requirements - maybe avoiding the need for the code using that call. Please check that if you can, or this may get held up waiting for someone else (us?) to check.

Please separate these *_version() define from the X509 definitions regardless. A separate check in lib-checks.m4 should solve the issue of LibreSSL differences mentioned above and also cope with any compatibility definitions OpenSSL may have been built with.

cotequeiroz replied:
I'll have to check the licensing terms, but it should not be dependend on what names are being used.
`OPENSSL_VERSION` is the new name of the flag passed to `OpenSSL_version()` to return the text variant of the version number and the release date. They were renamed from `SSLEAY_VERSION` and `SSLeay_version()`, respectively. See https://www.openssl.org/docs/man1.1.0/crypto/OpenSSL_version.html, and compare it to https://www.openssl.org/docs/man1.0.2/crypto/SSLeay_version.html.
I can add a check for `OpenSSL_version()` in `lib-checks.m4`, but, like I said, they were all implemented at the same time, so it will cause an unnecessary check when configuring.

cotequeiroz replied:
There was only one line using both of these symbols, so I can just check if `SSLEAY_VERSION` is defined. I've checked in libressl (2.6,2.7,2.8) and openssl (1.02, 1.1.0 and 1.1.1), and it is always defined as 0, so it should be safe to use it as a check for which version to use.

yadij commented on src/security/Session.h:
```diff
@@ -18,6 +18,31 @@
+#define X509_set1_notBefore X509_set_notBefore
+#define X509_set1_notAfter X509_set_notAfter
+#endif
+/// lhash.h functions, on the other hand, need a check of their own.
```
Please document this #if code block independent of anything else going on. As if it was the only thing being done. It should also go in security/forward.h. Entries in that file should be alphabetical order as much as possible, which would make this bit not only separately located to the earlier block, but also before it.

cotequeiroz replied:
This block is only needed in `src/security/cert_generators/file/certificate_db.cc`, so I moved it to `src/security/cert_generators/file/certificate_db.h`.

yadij commented on src/security/Session.h:
```diff
@@ -18,6 +18,31 @@
+#if !HAVE_LIBCRYPTO_X509_STORE_CTX_SET0_UNTRUSTED
+#define OPENSSL_VERSION SSLEAY_VERSION
+#define OpenSSL_version SSLeay_version
+#define X509_getm_notBefore X509_get_notBefore
```
These should probably be in security/forward.h. Please use alphabetical order (as much as possible anyway) for the definitions placed in there. Also, we prefer inline functions over #define where possible.

yadij commented on src/ssl/support.cc:
```diff
@@ -1172,7 +1195,11 @@ untrustedToStoreCtx_cb(X509_STORE_CTX *ctx,void *data)
     if (SquidUntrustedCerts.size() > 0)
         completeIssuers(ctx, sk);
 
+#if HAVE_LIBCRYPTO_X509_STORE_CTX_SET0_UNTRUSTED
```
This looks like it should be getting the same renaming + definition treatment the other calls are given.

I have not investigated whether these two calls are actually drop-in identical. That will need to be checked before approval of this PR. If they are not, then this is fine but the gain or behaviour tradeoff between the two needs to be documented (here or in commit message).

cotequeiroz replied:
It's easy to see they're drop-in identicals by looking at the openssl 1.1 headers.  What the do is add a define for the old name when previous-versions API compatibility is needed.  You could do it either by using the new name or the old one (and libressl is keeping both defined).  I would prefer to use the new name in the main code, and add a define for it when using na old version.  With time, the old version should be dropped, and the main code will not need changes.

yadij commented on src/security/cert_generators/file/security_file_certgen.cc:
```diff
@@ -324,9 +324,11 @@ int main(int argc, char *argv[])
             Ssl::CertificateDb::Check(db_path, max_db_size, fs_block_size);
         }
 
+#if HAVE_LIBSSL_SSL_LOAD_ERROR_STRINGS
```
This SSLeay_add_ssl_algorithms() is missing a check for itself. We cannot really guarantee that all possible environments will be lacking both. Better to check for both and wrap separately.

Same for the calls in ssl/support.cc

cotequeiroz replied:
Initialization is done automatically in openssl 1.1, so neither call needs to be made.  Openssl 1.1 has a replacement, which can be used in case of a non-standard initialization, which is not the case in squid.  There should be no need to call one or the other separately, so the extra check is not needed.

cotequeiroz commented:
@yadij 
Thanks for the review.  I changed most of the things you asked.  The change from SSLeay to OpenSSL should have no bearing over licensing, but, as an exception to using the new names, OPENSSL will only be used if SSLEAY is not defined.
Cheers

rousskov commented:
OK to test

rousskov commented:
I plan to review this later, after Amos is happy, and I have not read Amos comments yet, but please note that we should avoid sprinkling code with `#ifdefs`, especially high-level code. If the same action/concept is covered by two different functions/names (depending on the library version or flavor), then we should add a wrapper function (or equivalent) that higher level code can use unconditionally.

yadij commented:
If you wish. The convention we use for shim's etc would mean adding compat/openssl.h which includes openssl/ssl.h and provides shims for anything we need that is missing. Then include compat/openssl.h into compat/compat.h .

FWIW: the reason I have not added this already is that the OpenSSL API is spread over a number of openssl/*.h files which get included in different places and do not need to be included widely.

rousskov commented:
> The convention we use for shim's etc would mean adding compat/openssl.h which includes openssl/ssl.h and provides shims for anything we need that is missing.

Please do not misinterpret that statement as the recipe to blindly follow in all cases. You may add compat/openssl_foo.h or similar shims to properly group related OpenSSL shims together while isolating them from other OpenSSL shims. Adding 20 such foo.h files for OpenSSL would probably be a bad idea, but properly wrapping a few hand-picked OpenSSL headers may work well. Focus on the resulting code quality first, and then apply any conventions you know about, not the other way around.

Also, if a shim is too complex for `compat/` (e.g., it needs something from `src/`), then it may need to go into `src/ssl/`.

cotequeiroz commented:
I've created the shim, and moved most of the wrappers to `compat/openssl.h`.  I've made some necessary changes to fix compilation with some of the libraries.  Notably, the `EVP_PKEY_up_ref()` was not being correctly done, so even if this PR does not get merged, it may be worth fixing it.  Also, I've implemented the proposed fix (remove `const` from `useMethod` in `src/ssl/bio.cc`, so it would compile with libressl-2.7.  I've compiled tested it with openssl-1.0.0s, 1.0.1u, 1.0.2p, 1.1.0i, and 1.1.1-pre9, and with libressl-2.6.5, 2.7.4, and 2.8.0.
I'm not sure which versions of these are officially supported.  I've tried openssl-0.9.8zh, and it almost works as is--there's some parts of the code that use `SSL_OP_NO_TLSv1_{1,2}` without testing if they were defined first.
I haven't fixed this as I'm not sure what versions of which library are "officially"? supported, or even if we want to allow compilation with an unmaintained TLS library--I would guess not.

cotequeiroz commented on src/security/forward.h:
```diff
@@ -30,39 +36,6 @@
-#if !HAVE_LIBCRYPTO_EVP_PKEY_UP_REF
-#if defined(CRYPTO_LOCK_EVP_PKEY) // OpenSSL 1.0
-inline int EVP_PKEY_up_ref(EVP_PKEY *t) {if (t) CRYPTO_add(&t->references, 1, CRYPTO_LOCK_EVP_PKEY); return 0;}
-#endif
```
Here's where the EVP_PKEY_up_ref check is not right.  The `#endif` should probably be after the `#else` clause, like in previous checks.  Most importantly, though, there's no check for the function in `acinclude/libchecks.m4`, so the condition  (`!HAVE_LIBCRYPTO_EVP_PKEY_UP_REF`) is always true, and the error is never shown, which is certainly not the intention.

yadij replied:
ah, thank you.

cotequeiroz commented:
The build is failing Semaphore.  At first, it was complaining about missing function prototypes.  Well, they're all inline, and have `#if` wrappers around them.  I do no really see a good-looking solution.  We can put the prototype within the `#if`, like 
´´´
#if !HAVE_LIBCRYPTO_ASN1_STRING_GET0_DATA
inline const unsigned char *ASN1_STRING_get0_data(const ASN1_STRING *x);
inline const unsigned char *
ASN1_STRING_get0_data(const ASN1_STRING *x)
{
    return x->data;
}
#endif
´´´
Or we can put all prototypes at the top, all with their own, repeated, `#if` condition, which will not help readability anyway.
I just thought it would be easier just to add some `#pragma`s and silence the compiler instead.
Now it gives me some linking error, as if it's including code from `compat/openssl.h` where it shouldn't.

I did run into some errors like that when I was testing it here, but by conclusion at the time was that I was linking with different, incompatible, openssl versions, one being pulled from ldap, and another from squid itself, since it would not happen when I linked against the system's openssl.  I'll take a look at it.

cotequeiroz commented:
The error I got was very similar: 
```
make[5]: Entering directory '/home/equeiroz/src/squid.work/src/acl/external/LDAP_group'
depbase=`echo ext_ldap_group_acl.o | sed 's|[^/]*$|.deps/&|;s|\.o$||'`;\
g++ -DHAVE_CONFIG_H -DDEFAULT_CONFIG_FILE=\"/home/equeiroz/src/squid.work/openssl-1.0.0/etc/squid.conf\" -DDEFAULT_SQUID_DATA_DIR=\"/home/equeiroz/src/squid.work/openssl-1.0.0/share\" -DDEFAULT_SQUID_CONFIG_DIR=\"/home/equeiroz/src/squid.work/openssl-1.0.0/etc\"   -I../../../.. -I../../../../include -I../../../../lib -I../../../../src -I../../../../include    -I/home/equeiroz/src/squid.work/openssl-1.0.0/include -I/usr/include/libxml2 -DOPENSSL_NO_DEPRECATED -I/usr/include/libxml2 -Wall -Wpointer-arith -Wwrite-strings -Wcomments -Wshadow -Woverloaded-virtual -Werror -Wno-deprecated-register -pipe -D_REENTRANT -g -O2 -march=native -MT ext_ldap_group_acl.o -MD -MP -MF $depbase.Tpo -c -o ext_ldap_group_acl.o ext_ldap_group_acl.cc &&\
mv -f $depbase.Tpo $depbase.Po
/bin/sh ../../../../libtool  --tag=CXX   --mode=link g++ -I/usr/include/libxml2 -Wall -Wpointer-arith -Wwrite-strings -Wcomments -Wshadow -Woverloaded-virtual -Werror -Wno-deprecated-register -pipe -D_REENTRANT -g -O2 -march=native  -g -o ext_ldap_group_acl ext_ldap_group_acl.o ../../../../lib/libmiscencoding.la ../../../../compat/libcompatsquid.la  -lldap -llber -lm -lnsl -lbsd -lresolv -lcap -lnetfilter_conntrack -lrt
libtool: link: g++ -I/usr/include/libxml2 -Wall -Wpointer-arith -Wwrite-strings -Wcomments -Wshadow -Woverloaded-virtual -Werror -Wno-deprecated-register -pipe -D_REENTRANT -g -O2 -march=native -g -o ext_ldap_group_acl ext_ldap_group_acl.o  ../../../../lib/.libs/libmiscencoding.a ../../../../compat/.libs/libcompatsquid.a -lldap -llber -lm -lnsl -lbsd -lresolv -lcap /usr/lib64/libnetfilter_conntrack.so -L/usr/lib64 -lnfnetlink -lmnl -lrt
/usr/lib/gcc/x86_64-pc-linux-gnu/7.3.0/../../../../x86_64-pc-linux-gnu/bin/ld: ../../../../lib/.libs/libmiscencoding.a(rfc1738.o): undefined reference to symbol 'CRYPTO_add_lock'
/usr/lib64/libcrypto.so.1.0.0: error adding symbols: DSO missing from command line
collect2: error: ld returned 1 exit status
make[5]: *** [Makefile:824: ext_ldap_group_acl] Error 1
```
It would happen if I compiled it for, say, openssl 1.0.0, while my system was running 1.0.2--the so file is libcrypto.so.1.0.0 for both of them.
However, before the latest changes, Semaphore was building just fine, so I don't know what to think.

yadij commented:
The prototypes complaint is happening because the compat code is included in both C and C++ objects. This whole file should be wrapped in #if __cplusplus checks to make it empty for the C code and should not have any prototypes (which are only needed in C).

I'm not sure about the DSO. AIUI that should only appear if one of the inline is used and makes a symbol dependency. Lets see if it goes away with the above change.

rousskov commented:
I am concerned about the volume and complexity of checks/wrappers/workarounds being added in this PR. It feels like we may be going in the wrong overall direction here -- creating a significant long-term liability for ourselves while duplicating a lot of work already done by OpenSSL. I am not sure that support for OPENSSL_NO_DEPRECATED is worth all these changes. The current `saves some real estate` motivation in the PR description is not detailed/convincing enough IMO. Please detail it -- why is OPENSSL_NO_DEPRECATED support worth all this short-term work and long-term maintenance?

And, if OPENSSL_NO_DEPRECATED _is_ deemed worth all this trouble for some currently unknown (to me) reason, then, for Squid master/v5, we should at least consider dropping support for OpenSSL versions that require using those now-deprecated features. However, dropping support for older OpenSSL versions should only be discussed if/after OPENSSL_NO_DEPRECATED support is proven to be worth it in the first place.

cotequeiroz commented:
Perhaps I've got carried away and did too much with this PR.
My motivation was initially to avoid the use of the deprecated API to allow the openssl library to be built without the support for deprecated API, which makes the library about 20Kb smaller. While this might not be much, it is wasted space when you have to make the whole system fit in an 8MB flash.  In my use, anyway, I don't keep squid in flash, I use an external HD, but openssl is kept in flash, so the router can work without the HD.

The benefit of updating the API, for everybody else, is in keeping the main, high-level, code up-to-date; with time, it will be forced to get updated anyway.
There are my initial commit:
1 - header files:  when you include <openssl/ssl.h> without OPENSSL_NO_DEPRECATED, it loads many other headers, which were present in openssl for a long time.  Without that flag, the headers need to be included by themselves.  This should not be much of a burden.  All of the headers being included exist in any of the supported libraries.
2 - renamed functions: there were many calls that just got renamed.  The new versions keep the old names defined. With OPENSSL_NO_DEPRECATED, the old names are not defined and stop working.  So, my approach was to do it in reverse:  define the new names when you compile with the old library.  This way the main code stays up-to-date, and we deal with the renames in the compat header.  Note that some of this has already taken place when openssl 1.1 was initially supported--the api changed drastically between 1.0.2 and 1.1.  So what I've done was to move all of the compatibility shims in one place.
3 - Library initialization:  in openssl  < 1.1 (libressl < 2.7), you needed to make some calls to initialize the library before using it.  In later versions, this has been eliminated.  There are still initialization functions, so if, for example, you want to load the library without error message, you must call OPENSSL_init_crypto.  In squid, the error messages are always loaded, so the initialization call is not needed.  Again, there needs to be a check if the library needs to be initialized or not.
4 - There's a call to `RSA_generate_key` that needs to be changed to `RSA_generate_key_ex`.  This is the only actual API change that was done.  This call is part of the old export-ciphersuites key exchange that itself should not be needed.  It was removed in libressl, and openssl 1.1.0, and it doest not get compiled.  My recommendation would be to just drop the code guarded by `HAVE_LIBSSL_SSL_CTX_SET_TMP_RSA_CALLBACK`, so that the deprecated call would not even be there. I haven't tried to do this here.
While this is not the first time I've adapted code to openssl 1.1 API, this is my first contribution to squid, and so I'm not used to the way things work here, sorry.  While I still want this to get merged, I completely understand why you may not want to.
Cheers

cotequeiroz commented on src/ssl/bio.cc:
```diff
@@ -706,12 +706,7 @@ applyTlsDetailsToSSL(SSL *ssl, Security::TlsDetails::Pointer const &details, Ssl
-            const SSL_METHOD *method = SSLv23_method();
-            const SSL_CIPHER *c = method->get_cipher_by_char(cbytes);
-#endif
             if (c != NULL) {
```
Got it.

cotequeiroz commented on compat/openssl.h:
```diff
@@ -0,0 +1,210 @@
+EVP_PKEY_get0_RSA(EVP_PKEY *pkey)
+{
+    if (pkey->type != EVP_PKEY_RSA)
+        return NULL;
```
Done.

cotequeiroz commented on compat/openssl.h:
```diff
@@ -0,0 +1,210 @@
+#if !HAVE_LIBCRYPTO_X509_GET0_SIGNATURE
+inline void
+X509_get0_signature(ASN1_BIT_STRING **psig, X509_ALGOR **palg,
+                    const X509 *x)
```
Done.

yadij commented:
Updated the title and description to better match what is now being done, and to conform with wrap requirements of commit message.

yadij commented:
@rousskov, you marked yourself for a review. I am leaving the merge label and review marking for you to add and remove respectively.

rousskov commented:
Yes, the PR is still waiting for my review.

rousskov commented on src/security/cert_generators/file/certificate_db.h:
```diff
@@ -14,6 +14,9 @@
 #include <openssl/opensslv.h>
 #endif
+#if HAVE_OPENSSL_LHASH_H
+#include <openssl/lhash.h>
```
This addition puzzles me. Either it is not needed because `compat/openssl.h` includes it OR this file is used where `compat/openssl.h` itself is _not_ included.

cotequeiroz replied:
I think this was left over from before the shim was used.  I'll remove it.
BTW, `<openssl/opensslv.h>` could also be dropped here, as all it does is define `(OPENSSL|SHLIB)_VERSION_.*` macros, which are not being used anywhere this file is included.

cotequeiroz commented on compat/compat.h:
```diff
@@ -117,5 +117,21 @@
+ * Once all the supported library versions use the new API, the shim
+ * can be dropped.
+ */
+#include "compat/openssl.h"
```
I will find out which source files use openssl code and include de header there only.  It also avoids having to deal with the shim being included in C sources.  Right now, the header is empty when included by a C source, which just does not feel right.

cotequeiroz commented on compat/openssl.h:
```diff
@@ -0,0 +1,209 @@
+
+#if !HAVE_LIBCRYPTO_ASN1_STRING_GET0_DATA
+inline const unsigned char *
+ASN1_STRING_get0_data(const ASN1_STRING *x)
```
I will make those changes.

cotequeiroz commented on compat/openssl.h:
```diff
@@ -0,0 +1,209 @@
+}
+#endif
+
+// BIO_get_init is not implemented in LibreSSL, even though BIO_set_init is
```
OK

cotequeiroz commented on src/main.cc:
```diff
@@ -677,7 +677,11 @@ mainHandleCommandLineOption(const int optId, const char *optValue)
+#if defined(SSLEAY_VERSION)
         printf("\nThis binary uses %s. ", SSLeay_version(SSLEAY_VERSION));
+#else
+        printf("\nThis binary uses %s. ", OpenSSL_version(OPENSSL_VERSION));
```
My idea was to just use the new API in the high-level code, and deal with renames in the shim, but @yadij had concerns about licensing changes when using the new names.  That's why I changed it to use the `#if`.  I can't see how the use of new function names can even affect licensing in this case, after reading the simple openssl license file, and particularly since the source can be compiled with a different library anyway. However, I'm not a lawyer.  Please advise me how to deal with this.  If I can have a vote, it would be to use the new API and add use the `#if` in the shim.

yadij replied:
The TLD;DR answer is yes a shim is fine.

Changing this line does not change the license, quite the reverse. This line is informing about the license conditions imposed from old SSLEAY and OpenSSL versions copyright.

OpenSSL project are on the cusp of changing their license and I was expecting the SSLEAY_VERSION to be removed in versions where SSLEAY no longer applies. OPENSSL_VERSION is expected to remain in all OpenSSL builds regardless of license.

For our part we only need this text where the old SSLEAY legal situation applies to the library. So I was asking about a check primarily to see whether we could use that version macro as indicator of that state - the answer to that was no. Unfortunately for us the OpenSSL versions still currently have the old OpenSSL-specific clauses even when the SSLEAY clauses do not apply.

cotequeiroz replied:
I'm using the new API for these calls.  I meant to have done it already, but I missed it, sorry.

cotequeiroz commented on compat/openssl.h:
```diff
@@ -0,0 +1,209 @@
+EVP_PKEY_up_ref(EVP_PKEY *t)
+{
+    if (t)
+        CRYPTO_add(&t->references, 1, CRYPTO_LOCK_EVP_PKEY);
```
This function is not implemented correctly.  I can't return `CRYPTO_add` directly, since it returns the updated reference count, while `EVP_PKEY_up_ref` should return 1 for success and 0 for failure.  As a note, we were always signaling failure; apparently, the return values are never checked.  I will fix all of the similar `.*_up_ref` functions that have this same behaviour, using the same logic as openssl > 1.1 does, returning 1 if the updated reference count > 1, and 0 otherwise.

cotequeiroz commented on src/ssl/support.cc:
```diff
@@ -97,18 +97,34 @@ ssl_ask_password(SSL_CTX * context, const char * prompt)
+        e = BN_new();
+        if (!e || !BN_set_word(e, RSA_F4)) {
+            debugs(83, DBG_IMPORTANT, "ssl_temp_rsa_cb: Failed to set exponent for key " << keylen);
+            BN_free(e);
```
This is documented behavior.

cotequeiroz commented on src/ssl/support.cc:
```diff
@@ -117,26 +133,31 @@ ssl_temp_rsa_cb(SSL * ssl, int anInt, int keylen)
     default:
         debugs(83, DBG_IMPORTANT, "ssl_temp_rsa_cb: Unexpected key length " << keylen);
-        return NULL;
+        return nullptr;
```
My motivation to change these in particular was just to maintain consistency within the function.  What I would suggest is just to change the variable declarations block to `nullptr`, and undo the rest.  It looks odd to have just one variable different than the others.

cotequeiroz commented on src/ssl/gadgets.cc:
```diff
@@ -948,7 +943,6 @@ Ssl::CertificatesCmp(const Security::CertPointer &cert1, const Security::CertPoi
 Ssl::X509_get_signature(const Security::CertPointer &cert)
 {
-#if HAVE_LIBCRYPTO_X509_GET0_SIGNATURE
 #if SQUID_USE_CONST_X509_GET0_SIGNATURE_ARGS
```
I gave this a lot of thought.  This function is already a wrapper to X509_get0_signature, whose sole purpose is to "[take] care of portability issues with older OpenSSL versions".  So, to avoid creating two wrappers, I've moved the function to compat/openssl.h, adapting the argument type to avoid needing to depend on other headers.  However, I think a simpler approach is just to use the `SQUID_CONST_X509_GET_SIGNATURE_ARGS` macro, either defined as `const` or empty, to handle the arguments type.  I'm leaving the commits for the individual options, so you can see what the code would be like.

cotequeiroz commented:
I've push commits with most of the changes you requested.  I'll squash them if they're approved.  I'm doing the first request--to include the shim only where it is needed--last, as it is not really straightforward for me; I will study the squid code more before I add more commits.

cotequeiroz commented:
@rousskov 
See if these commits achieve what you asked me to do, and let me know.  The last commit was beyond what you asked me, but it is part of the `#if ... #else ... #endif` reduction related to openssl, and I think it belongs in the shim as well.

rousskov commented:
Thank you for moving this PR forward! Please give me a few days to re-review this. Please do not squash the commit manually -- [Anubis](https://github.com/measurement-factory/anubis) will do that automatically when merging.

rousskov commented:
Please review the updated PR description. It will become the commit message body when the changes are merged by Anubis. I edited it to remove LibreSSL reference and minor details that are already reflected in the diff and are no longer the focus of the PR.

rousskov commented:
@chtsanti, please take a high-level look at these changes if you get a chance. This PR moves/enhances a lot of code that you worked on. You may be able to spot something that others have missed.

cotequeiroz commented on compat/openssl.h:
```diff
@@ -0,0 +1,241 @@
+#error compat/openssl.h depends on USE_OPENSSL
+#endif
+
+#if defined(__cplusplus)
```
The two conditions above are true.

The code is not compatible with C as is.  Besides the use of a couple of `nullptr` in the code (OK, this is trivial), there is the difference in semantics for `inline` between C89 (no inline at all), GNU89, C99 and C++ that's a bit awkward to accommodate.  If you leave it as is and compile it under C99 rules, the compiler will generate warnings about missing declarations, but should work with optimization.  It took me a while to understand what was going on.  To make it portable, an `extern inline` definition must exist somewhere, so that the function can be called if it is not actually inlined.  

What I think it's the best solution is to use `defined(__cplusplus)` to assert that `compat/openssl.h` is being used only in C++.  If there is ever a need to use it in C, then the extern inline definitions can be created.  I'll add a note to the assertion as a guide.

cotequeiroz replied:
Got it.

cotequeiroz commented on src/ssl/ErrorDetail.cc:
```diff
@@ -490,7 +490,7 @@ const char *Ssl::ErrorDetail::ca_name() const
 {
     if (broken_cert.get()) {
-        if (ASN1_UTCTIME * tm = X509_get_notBefore(broken_cert.get())) {
+        if (ASN1_UTCTIME * tm = X509_getm_notBefore(broken_cert.get())) {
```
I'll do it.

cotequeiroz commented on src/ssl/ErrorDetail.cc:
```diff
@@ -505,7 +505,7 @@ const char *Ssl::ErrorDetail::notbefore() const
 {
     if (broken_cert.get()) {
-        if (ASN1_UTCTIME * tm = X509_get_notAfter(broken_cert.get())) {
+        if (ASN1_UTCTIME * tm = X509_getm_notAfter(broken_cert.get())) {
```
Ditto.

rousskov commented on src/security/ServerOptions.cc:
```diff
@@ -17,13 +17,10 @@
 #include "SquidConfig.h"
 #if USE_OPENSSL
 #include "ssl/support.h"
```
Thank you. 

[rant]You cannot really know for sure whether `openssl/err.h` needs our shim because both may change outside your control. It would be nice to be able to rely on the current state of code (and on folks carefully checking all side effects of their future changes), but that virtually never works long-term. Thus, when we can, we should use future-proof rules that are easy to apply/review and are programmatically verifiable. We are lucky that the rule we (mostly you) came up with during this PR is such a rule.[/rant]

cotequeiroz commented on compat/openssl.h:
```diff
@@ -0,0 +1,241 @@
+#define OPENSSL_LH_strhash lh_strhash
+#endif
+
+inline void SQUID_OPENSSL_init_ssl(void)
```
I've ran the script, and it changed the indentation of the shim functions to conform to the `extern "C"` braces.  Since I found more code that followed this style, I've committed the change.

rousskov replied:
I think you did the right thing here: We should accept astyle imperfect formatting -- tuning it is outside this PR scope. It is perhaps debatable whether we should have such a huge `extern "C"` block, but that is a separate question (that I am not asking on this change thread).

rousskov commented on compat/openssl.h:
```diff
@@ -0,0 +1,248 @@
+#endif
+
+    inline void
+    SQUID_OPENSSL_init_ssl(void)
```
Since this is a Squid-only function, not an OpenSSL function wrapper, it probably should not be inside extern "C", but this is not a big deal.

cotequeiroz replied:
I was about to comment on this after your previous post, but you beat me to it.  Do you want me to leave it out the the `extern "C"` block?

rousskov replied:
>  Do you want me to leave it out the the `extern "C"` block?

Yes, but I do not insist on this change.

> where should it be placed?

Personally, I would put it _after_ the block so that other (future) Squid-only functions (if any) placed near it can benefit  from the wrappers above them. Above is also OK (and even natural).

I would not split the block -- all those wrapping/shimming functions are logically in the same category (in relation to Squid-only functions) so it is probably best to keep them grouped until we find a valid reason to separate them.

cotequeiroz replied:
Done.

cotequeiroz commented:
I've reviewed the description as you requested.  Everything OK here.

rousskov commented:
> cotequeiroz added a commit to cotequeiroz/squid that referenced this pull request 19 hours ago

FYI: Anubis ignores individual PR commit messages. To create the final to-be-merged commit [message](https://github.com/measurement-factory/anubis#commit-message), Anubis uses the PR title and description instead.

rousskov commented:
@yadij, I have not labeled this PR for merging to give you a chance to review the changes that happened after your last review. I will label it in a day or so unless you stop me. Please feel free to do that yourself at any time, of course.

chtsanti commented:
> @chtsanti, please take a high-level look at these changes if you get a chance. This PR moves/enhances a lot of code that you worked on. You may be able to spot something that others have missed.

Looks OK.

rousskov commented:
@cotequeiroz, any ideas how this PR can cause Squid bug [4918](https://bugs.squid-cache.org/show_bug.cgi?id=4918#c11)?

cotequeiroz commented:
My bet is that the initialization is not being done properly (`./configure` detection failing?).  I'll check old API, run some tests here, and will get back to you either here or at the bug report (tell me if you have a preference).

rousskov commented:
> I'll check old API, run some tests here, and will get back to you either here or at the bug report (tell me if you have a preference).

Thank you! If you discover more information about the bug, please add that info to the bug [report](https://bugs.squid-cache.org/show_bug.cgi?id=4918). If you have a fix, please open a new PR (and add a link to it to the bug report). I do not think we should update this closed PR.
